proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+      result = pipe.run(query=query, questions={"bind_result": "MultiTaskComponent_0"})
+
+      print(result["results"])
+  ```
+
+  In the example above we show another way to bind the results of one MultiTaskComponent to the input of another by
+  using keyword arguments. The `questions` keyword parameter is used to bind the results of the `MultiTaskComponent_0`
+  node and will be available to all downstream nodes via `invocation context` (hidden inside MultiTaskPipeline).
+
+
+  - Default tasks/prompts to be added to MultiTaskComponent
+
+    [Muffin]:
+      - Summarization
+      - Natural Language Inference
+      - Multiple-Choice QA
+      - Translation
+      - Sentiment Analysis
+      - Extractive QA
+      - Structured Data to Text
+      - Coreference Resolution
+      - Code Repair
+      - Code Error Generation
+      - Dialogue Context Generation
+      - Closed-Book QA
+      - Next Sentence Prediction
+      - Paraphrasing Identification
+      - Conversational Question Answering
+      - Topic Classification
+      - Mathematical QA
+      - Dialog Next Turn Prediction
+      - Grammatical Acceptability
+      - Punctuation fixing
+
+    [T0-SF]:
+      - Adversarial QA
+      - Question Generation
+      - Commonsense Reasoning
+      - Title Generation
+      - Dialogue Turn Prediction
+      - Predict Span Indices
+      - Context Generation
+
+    [NIV2]:
+      - Program Execution
+      - Text Matching
+      - Toxic Language Detection
+      - Cause Effect Classification
+      - Information Extraction
+      - Textual Entailment
+      - Wrong Candidate Generation
+      - Named Entity Recognition
+      - Commonsense Classification
+      - Fill-in-the-blank
+      - Text Completion
+      - Sentence Composition
+      - Question Understanding
+
+    [CoT Reasoning]:
+      - Explanation Generation
+      - Generate Question And Answer
+      - Grade School Math Word Problems
+      - Algebraic Question Answering
+      - Common Sense Reasoning Over Entities
+      - Common Sense Reasoning For QA
+      - Passage Based Question Answering
+      - Sense-Making And Explanation
+
+  # Drawbacks
+- One potential drawback of MultiTaskComponent is that it may require a significant amount of computational resources
+  to use. This may limit its use in applications or environments where there are constraints on the available hardware
+  or software resources. In addition, MultiTaskComponent may not be suitable for all NLP tasks or use cases, and users
+  may need to evaluate its performance and suitability for their specific needs and requirements.
+
+
+- During the prototyping phase we found that in order to follow current pipeline design patterns we needed to make
+  certain compromises. If a pipeline has more than one MultiTaskComponent, we can not simply use one instance of
+  MultiTaskComponent and switch the tasks in runtime. Instead, we need to create a new instance of MultiTaskComponent
```

Comment:
That does sound like a real bummer. What is causing this issue with the current pipeline design?

Reply:
Can't use arbitrary object graph to create pipeline nodes unless they are part of the pipeline as well. Think how retriever can take doc store in its init method but doc store isn't some independent component created outside of pipeline context

Reply:
Definitely something to tackle in future, but nothing I would put in scope for the initial version.

Reply:
Thank you both - keeping this unresolved so others can see this comment about the current downside!

Reply:
I'd also say that this limitation might not be as big in practice. If you use a true foundation model for the `MultiTaskComponent`, then I'd say that you would probably host this model as an API anyways and just send requests with the different prompts from the component instances. So it wouldn't mean that you'd have to keep such a big model in memory multiple times.